### ISSUE TITLE
DevTrace and Turbolinks

### ISSUE THREAD

**Issue by fredoliveira** [scoutapp/scout_apm_ruby#97]
Hey guys,

Superb job on Scout and DevTrace. One snag I found is that DevTrace doesn't play well with Turbolinks (tested on a Rails 5.1 app). Because of the way turbolinks works (effectively merging the head and replacing the body on the fly on new requests), the devtrace widget cannot be found after the first request.

```
Uncaught TypeError: Cannot read property 'unique_id' of undefined
    at expandedProfileToHTML (scout_instant.js?cachebust=1496923205:224)
    at ub.<anonymous> (scout_instant.js?cachebust=1496923205:161)
    at p (scout_instant.js?cachebust=1496922257:4)
    at ub.each (scout_instant.js?cachebust=1496922257:4)
    at renderMessages (scout_instant.js?cachebust=1496923205:155)
    at initialize (scout_instant.js?cachebust=1496923205:95)
```

I have debugged this situation with other pieces of software before (namely Zendesk) and it usually requires a bit of refactoring so that the widget rebuilds itself on a `turbolinks:load` event (their documentation on that is [available here](https://github.com/turbolinks/turbolinks#working-with-script-elements)). 

Happy to provide further information on what's happening, if I'm able.

**Comment by alfie-max** [comment]
Was about to create a new issue when i saw this. Guys can we look into this issue soon?

**Comment by itsderek23** [comment]
Sorry for the delay on this - when I test with:

* Rails 5.1.1
* turbolinks 5.0.1
* scout_apm 2.1.24
* chrome 58.0.3029.81

...I'm seeing the speed badge update as I navigate. It appears the requests are XHR. Are you seeing this in every case, or just specific cases?

**Comment by Chadh13** [comment]
![web image 2018-06-01 16-59-28](https://user-images.githubusercontent.com/2566348/40865284-2a8a1b14-65bd-11e8-9341-c6f42b0c52af.jpg)
![ssa 2018-06-01 17-11-19](https://user-images.githubusercontent.com/2566348/40865617-d83c5186-65be-11e8-8259-06777089f562.jpg)

Encountered this today *only after* I switched over one of our apps to `'data-turbolinks-eval' => true` for all JS within our `application.js` file.

NOTE: `http://localhost:3000/mentors/000` loads the scout UI and resources fine, but clicking down into `http://localhost:3000/mentors/000/students/000` would not. After I opened the UI to see the performance stats on the page it DID work on, I couldn't seem to reproduce this behavior any longer after that.

Here's my system:
- Rails 5.0.0
- turbolinks 5.0.1
- scout_apm 2.4.12
- chrome 66.0.3359.181

and here are some logs that may or may not be helpful:
```
Started GET "/mentors/114/students/addTrace/instant/minimize.png" for ::1 at 2018-06-01 17:10:05 -0500

ActionController::RoutingError (No route matches [GET] "/mentors/114/students/addTrace/instant/minimize.png"):

actionpack (5.0.0) lib/action_dispatch/middleware/debug_exceptions.rb:53:in `call'
actionpack (5.0.0) lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
railties (5.0.0) lib/rails/rack/logger.rb:36:in `call_app'
railties (5.0.0) lib/rails/rack/logger.rb:24:in `block in call'
activesupport (5.0.0) lib/active_support/tagged_logging.rb:70:in `block in tagged'
activesupport (5.0.0) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (5.0.0) lib/active_support/tagged_logging.rb:70:in `tagged'
railties (5.0.0) lib/rails/rack/logger.rb:24:in `call'
sprockets-rails (3.2.1) lib/sprockets/rails/quiet_assets.rb:13:in `call'
request_store (1.4.0) lib/request_store/middleware.rb:19:in `call'
actionpack (5.0.0) lib/action_dispatch/middleware/request_id.rb:24:in `call'
rack (2.0.3) lib/rack/method_override.rb:22:in `call'
rack (2.0.3) lib/rack/runtime.rb:22:in `call'
actionpack (5.0.0) lib/action_dispatch/middleware/executor.rb:12:in `call'
actionpack (5.0.0) lib/action_dispatch/middleware/static.rb:136:in `call'
rack (2.0.3) lib/rack/sendfile.rb:111:in `call'
rack-mini-profiler (0.10.5) lib/mini_profiler/profiler.rb:282:in `call'
scout_apm (2.4.12) lib/scout_apm/instruments/middleware_summary.rb:58:in `call'
railties (5.0.0) lib/rails/engine.rb:522:in `call'
puma (3.10.0) lib/puma/configuration.rb:225:in `call'
puma (3.10.0) lib/puma/server.rb:605:in `handle_request'
puma (3.10.0) lib/puma/server.rb:437:in `process_client'
puma (3.10.0) lib/puma/server.rb:301:in `block in run'
puma (3.10.0) lib/puma/thread_pool.rb:120:in `block in spawn_thread'
puma (3.10.0) lib/puma/server.rb:605:in `handle_request'
puma (3.10.0) lib/puma/server.rb:437:in `process_client'
puma (3.10.0) lib/puma/server.rb:301:in `block in run'
puma (3.10.0) lib/puma/thread_pool.rb:120:in `block in spawn_thread'
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb (2.6ms)
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb (1.2ms)
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb (1.3ms)
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_trace.html.erb (1.5ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [191 times] (3506.3ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [191 times] (3173.6ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [13 times] (3.7ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [13 times] (4.1ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [5 times] (1.5ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [5 times] (1.7ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_table.html.erb (3.1ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_table.html.erb (1.6ms)
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (1.1ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout (50165.7ms)
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (2.0ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout (51035.3ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [191 times] (1762.4ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [13 times] (117.1ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [5 times] (1.4ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_table.html.erb (2.2ms)
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (1.2ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout (36014.8ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [191 times] (555.7ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [13 times] (15.3ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [5 times] (2.1ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_table.html.erb (220.0ms)
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (1.3ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout (35369.1ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [191 times] (942.5ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [13 times] (3.0ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [5 times] (1.3ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_table.html.erb (1.2ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [191 times] (400.2ms)
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [13 times] (519.8ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (0.9ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout (42532.3ms)
  Rendered collection of /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_route.html.erb [5 times] (1.4ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/routes/_table.html.erb (1.3ms)
  Rendering /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/_request_and_response.html.erb (2.1ms)
  Rendered /Users/chadheinen/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0/lib/action_dispatch/middleware/templates/rescues/routing_error.html.erb within rescues/layout (38283.5ms)
```

Ping me in slack if you'd like for me to edit this to include more details!

**Comment by mjsommer** [comment]
Having this same issue ... Has it ever been resolved?

Current workaround, is to remove Turbolinks.

**Comment by itsderek23** [comment]
Thanks @mjsommer for reporting your experience. 

We're still having trouble reproducing this on our end.

**Comment by mjsommer** [comment]
Thanks Derek, so my app (Rails 4) with straight Bootstrap runs ok with Scout, but when I start generating DataTables populated by Ajax, Scout breaks. Try creating a multi-column page, where the first column (DataTable A), creates and populates the second column (DataTable B). Chances are, you will see that the Scout Badge is there, but when you clink on it, the icons are broken, and the data is inaccessible.

I'd share my work, but is a startup idea that I'm working on. :)

**Comment by benhill** [comment]
I'd just like to chime in that I am experiencing the same problem.

**Comment by kylekeesling** [comment]
Rails 5.2.1
Ruby 2.5.3
scout_apm 2.4.20
Chrome 70.0.3538.110


Experiencing this as well - logs report `RoutingError`s for:
`/addTrace/instant/logo.png`
`/addTrace/instant/minimize.png`

and I get a javascript error in my console here:

**Comment by vesapalmu** [comment]
Same problem here: Same error, same environment, except scout_apm 2.6.3.
